### PR TITLE
Kusto Retry Logic : Retries on specific exception types

### PR DESCRIPTION
**Pull request by ShekharGupta1988**
This PR Contains:
- Kusto Retry Logic for specific Exception Types (governed by configuration)
- Using Backup cluster for Last Retry attempt (governed by configuration)
           - For instance, if the original query is targeted towards follower cluster, the final retry will happen against main cluster and vice-versa
           - for Partners who doesnt have backup clusters, the final retry will be against the same cluster.
- This logic will be automatically applied to `ExecuteQuery` and `ExecuteClusterQuery` but not to `Heartbeat Logic`
- To begin with, we are only retrying in case of :
              - Exceptions of type `WeakConsistencyEntityNotFoundException` or `InternalServiceError (520-UnknownError)` . In Last 7 Days, there are 3085 exceptions with these errors and P90 is 8.5 sec.
              - Only those kusto requests that failed within 10 seconds

- please note : I was not able to use Generic Retry logic we have since there were special handling required for kusto retry like change the cluster for last retry and so on.

### CODE REVIEW

File: src/Diagnostics.DataProviders/KustoSDKClient.cs [hunk 1]
```diff
@@ -117,7 +129,86 @@ public async Task<DataTable> ExecuteQueryAsync(string query, string cluster, str
...
+                    }
+
+                    attempt++;
+                }
+
+                if (attempt <= _config.MaxRetryCount) await Task.Delay(_config.RetryDelayInSeconds * 1000);
+
+            } while (attempt <= _config.MaxRetryCount);
+
+            if(executeQueryTask.IsCompleted && !executeQueryTask.IsFaulted && !executeQueryTask.IsCanceled)
```
Comment (puneetg1983) [inline src/Diagnostics.DataProviders/KustoSDKClient.cs:206]: >executeQueryTask.IsCompleted && !executeQueryTask.IsFaulted && !executeQueryTask.IsCanceled [](start = 15, length = 91)

You can use Task.IsCompletedSuccessfully
[https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.iscompletedsuccessfully?view=netcore-3.1](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.iscompletedsuccessfully?view=netcore-3.1)

File: src/Diagnostics.DataProviders/KustoSDKClient.cs [hunk 2]
```diff
@@ -117,7 +129,86 @@ public async Task<DataTable> ExecuteQueryAsync(string query, string cluster, str
...
+                            exceptionDetails
+                            );
+
+                    // Logic to check if retry needs to continue
+                    if(totalResponseTime.TotalSeconds > _config.MaxFailureResponseTimeInSecondsForRetry || !IsExceptionRetryable(attemptException))
+                    {
+                        DiagnosticsETWProvider.Instance.LogRetryAttemptMessage(
+                        requestId ?? string.Empty,
+                        source,
+                        $"Not continuing Retries after Attempt : {attempt}, Cluster: {cluster}, Query : {query}"
```
Comment (puneetg1983) [inline src/Diagnostics.DataProviders/KustoSDKClient.cs:193]: >Not continuing Retries [](start = 26, length = 22)

Should we also log totalResponseTime.TotalSeconds ?

File: src/Diagnostics.DataProviders/KustoSDKClient.cs [hunk 3]
```diff
@@ -15,48 +14,61 @@
...
         private string _clientId;
         private string _aadAuthority;
         private static ConcurrentDictionary<Tuple<string, string>, ICslQueryProvider> QueryProviderMapping;
+        private static List<string> exceptionsToRetryFor = new List<string>();
 
         /// <summary>
         /// Failover cluster mapping
         /// </summary>
-        public ConcurrentDictionary<string, string> FailoverClusterMapping { get; set; }
+        private ConcurrentDictionary<string, string> FailoverClusterMapping { get; set; }
```
Comment (puneetg1983) [inline src/Diagnostics.DataProviders/KustoSDKClient.cs:36]: >private [](start = 8, length = 7)

Is this property not used anywhere else in the project?

File: src/Diagnostics.DataProviders/KustoSDKClient.cs (hunk 2 above)
Comment (ShekharGupta1988) [reply src/Diagnostics.DataProviders/KustoSDKClient.cs:193]: Its already logged as part of attempt summary event just above.  But i can add it to this event as well

Comment (puneetg1983) [reply src/Diagnostics.DataProviders/KustoSDKClient.cs:193]: Never mind, I see it is logged above...

---
In reply to: [434590128](https://github.com/Azure/Azure-AppServices-Diagnostics/pull/514#discussion_r434590128) [](ancestors = 434590128)

File: src/Diagnostics.DataProviders/KustoSDKClient.cs (hunk 3 above)
Comment (ShekharGupta1988) [reply src/Diagnostics.DataProviders/KustoSDKClient.cs:36]: Yup. This shouldnt be public

File: src/Diagnostics.DataProviders/KustoSDKClient.cs [hunk 4]
```diff
@@ -15,48 +14,61 @@
...
             _requestId = requestId;
             _kustoApiQueryEndpoint = config.KustoApiEndpoint + ":443";
             _appKey = config.AppKey;
             _clientId = config.ClientId;
             _aadAuthority = config.AADAuthority;
             FailoverClusterMapping = config.FailoverClusterNameCollection;
+
+            if (!string.IsNullOrWhiteSpace(_config.ExceptionsToRetryFor))
+            {
+                exceptionsToRetryFor = _config.ExceptionsToRetryFor.Split(DataProviderConstants.CommonSeparationChars).Select(p => p.Trim()).ToList();
```
Comment (SharmaAjay19) [inline src/Diagnostics.DataProviders/KustoSDKClient.cs:55]: I like this split.

File: src/Diagnostics.DataProviders/KustoSDKClient.cs [hunk 5]
```diff
@@ -117,7 +129,86 @@ public async Task<DataTable> ExecuteQueryAsync(string query, string cluster, str
...
+                            $"Attempt : {attempt},  IsSuccessful : {attemptException == null}, Cluster: {cluster}, Query : {query}",
+                            Convert.ToInt64(totalResponseTime.TotalMilliseconds),
+                            invocationStartTime.ToString("HH:mm:ss.fff"),
+                            invocationEndTime.ToString("HH:mm:ss.fff"),
+                            exceptionType,
+                            exceptionDetails
+                            );
+
+                    // Logic to check if retry needs to continue
+                    if(totalResponseTime.TotalSeconds > _config.MaxFailureResponseTimeInSecondsForRetry || !IsExceptionRetryable(attemptException))
```
Comment (cindylovescoding) [inline src/Diagnostics.DataProviders/KustoSDKClient.cs:188]: Do we want to check if "IsSuccessful" here too?  If the retry succeed, we don't need to continue here as well right?

Comment (ShekharGupta1988) [reply src/Diagnostics.DataProviders/KustoSDKClient.cs:188]: Yup. The IsExceptionRetryable() takes care of that :)

Comment (cindylovescoding) [reply src/Diagnostics.DataProviders/KustoSDKClient.cs:188]: Oh I see. It's checking null exception. Got it!

**Pull request merged**